File: server/core/src/main/java/de/codecentric/iot/rapiro/blazeds/actors/BlazeDsPublishingActor.java
package de.codecentric.iot.rapiro.blazeds.actors;

import akka.japi.pf.ReceiveBuilder;
import akka.stream.actor.AbstractActorSubscriber;
import akka.stream.actor.ActorSubscriberMessage;
import akka.stream.actor.OneByOneRequestStrategy;
import akka.stream.actor.RequestStrategy;
import flex.messaging.Destination;
import flex.messaging.MessageBroker;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.flex.messaging.MessageTemplate;

/**
 * Created by christoferdutz on 18.10.16.
 */
public abstract class BlazeDsPublishingActor extends AbstractActorSubscriber implements InitializingBean {

    private static final Logger LOG = LoggerFactory.getLogger(BlazeDsPublishingActor.class);

    @Autowired
    private MessageBroker broker;

    @Autowired
    private MessageTemplate template;

    private String destinationName;

    public BlazeDsPublishingActor(String destinationName) {
        this.destinationName = destinationName;
        receive(ReceiveBuilder.
                match(ActorSubscriberMessage.OnNext.class, on -> on.element() instanceof Object,
                        onNext -> {
                            Object object = onNext.element();
                            LOG.debug("Sending " + object + " to destination " + destinationName);
                            template.send(destinationName, object);
                        }).
                build());
    }

    @Override
    public void afterPropertiesSet() throws Exception {
        flex.messaging.services.Service service = broker.getService("messaging-service");
        Destination destination = service.createDestination(destinationName);
        service.addDestination(destination);
        if(!service.isStarted()) {
            service.start();
        } else if(!destination.isStarted()) {
            destination.start();
        }
    }

    @Override
    public RequestStrategy requestStrategy() {
        return OneByOneRequestStrategy.getInstance();
    }

}

File: server/core/src/main/scala/de/codecentric/iot/rapiro/blazeds/actors/BlazeDsPublishingActor.scala
package de.codecentric.iot.rapiro.blazeds.actors

import akka.actor.Actor
import com.typesafe.scalalogging.LazyLogging
import de.codecentric.iot.rapiro.akka.ItemEvent
import flex.messaging.{Destination, MessageBroker}
import org.springframework.beans.factory.InitializingBean
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.flex.messaging.MessageTemplate

/**
  * Created by christoferdutz on 24.10.16.
  */
abstract class BlazeDsPublishingActor extends Actor with LazyLogging with InitializingBean {

  @Autowired private val broker: MessageBroker = null
  @Autowired private val template: MessageTemplate = null

  def destinationName:String

  override def receive: Receive = {
    case eventItem:ItemEvent[Object] =>
      try {
        template.send(destinationName, eventItem.getItem)
      } catch {
        case e: Exception => logger.error("An error occurred while processing incoming event {}", eventItem, e)
      }
  }

  override def afterPropertiesSet(): Unit = {
    val service = broker.getService("messaging-service")
    if (service.getDestination(destinationName) == null) {
      val destination: Destination = service.createDestination(destinationName)
      service.addDestination(destination)
      if (!service.isStarted) service.start()
      else if (!destination.isStarted) destination.start()
      logger.info("Initialized messaging destination: '{}'", destinationName)
    }
  }

}
